File: templates/problems/17_math_geometry/04_happy_number/typescript/solution.ts
/**
 * Use Floyd's cycle detection. O(log n) time, O(1) space.
 */
function sumDigitSquares(num: number): number {
    let sum = 0;
    while (num > 0) {
        const d = num % 10;
        sum += d * d;
        num = Math.floor(num / 10);
    }
    return sum;
}

export function solveIsHappy(n: number): boolean {
    let slow = n, fast = sumDigitSquares(n);
    while (fast !== 1 && slow !== fast) {
        slow = sumDigitSquares(slow);
        fast = sumDigitSquares(sumDigitSquares(fast));
    }
    return fast === 1;
}

File: templates/problems/17_math_geometry/04_happy_number/python/solution.py
def solve_is_happy(n: int) -> bool:
    """Use Floyd's cycle detection. O(log n) time, O(1) space."""

    def sum_digit_squares(num: int) -> int:
        s = 0
        while num > 0:
            d = num % 10
            s += d * d
            num //= 10
        return s

    slow, fast = n, sum_digit_squares(n)
    while fast != 1 and slow != fast:
        slow = sum_digit_squares(slow)
        fast = sum_digit_squares(sum_digit_squares(fast))
    return fast == 1
